### ISSUE TITLE
Problem with mesh refinement in 3d codim 1 case

### ISSUE THREAD

**Issue by starki0815** [dealii/dealii#7761]
This little program messes up the mesh in 3d codim 1:
```cpp
#include <iostream>
#include <fstream>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/grid/grid_out.h>

using namespace std;
using namespace dealii;

int main(){

 	//triangulation is a single 2d square element embedded into 3d space
	Triangulation<2, 3> plane;
 	Point<3> p1(1.,0.,0.);
 	Point<3> p2(1.,1.,0.);
 	Point<3> p3(1.,0.,1.);
 	Point<3> p4(1.,1.,1.);
 	vector<Point<3>> vertices={p1, p2, p3, p4};
 	GridGenerator::general_cell(plane,vertices);

 	//set the material id of the cell to 1
 	plane.begin_active()->set_material_id(1);

 	//now associate a spherical manifold object with manifold_id 1
 	//(this object isn't used in this minimal example, but there could
 	//in principle be other cells to which the manifold_id 1 is assigned)
 	SphericalManifold<2, 3> spherical_manifold;
 	plane.set_manifold(1, spherical_manifold);

 	//now refine twice (the second refinement messes up the mesh)
 	plane.refine_global(2);

 	//output the mesh
 	ofstream out("plane.vtk");
	GridOut gridOut;
	gridOut.write_vtk(plane, out);
}
```

It seems to me that the following lines in the relevant `execute_refinement` function cause the problem:
```cpp
                              if (spacedim > dim)
                                {
                                  if (line->at_boundary())
                                    // if possible honor boundary
                                    // indicator
                                    line->set_user_index(line->boundary_id());
                                  else
                                    // otherwise take manifold
                                    // description from the adjacent
                                    // cell
                                    line->set_user_index(cell->material_id());
                                }
                            }
```
In these code lines the `user_index` of the lines to be refined is set. Later in the code the `user_index` is used to determine the manifold object to be used for the new point on the line in case the line has `manifold_id==numbers::flat_manifold_id`:
```cpp
                if (line->manifold_id() == numbers::flat_manifold_id)
                  triangulation.vertices[next_unused_vertex] =
                    triangulation.get_manifold(line->user_index())
                      .get_new_point_on_line(line);
```
To me this is everything a little confusing, because I thought the manifold object to be used is solely determined by the `manifold_id` of the respective object. But here the `boundary_id` or the `material_id` are used if `manifold_id==numbers::flat_manifold_id`. Is there a reason for doing it this way or is it a bug? If it's done intentionally, shouldn't it be documented e.g. in the glossary (at least I wasn't able to find the place, where this behavior is documented)?

**Comment by masterleinad** [comment]
Can you show us an image of the resulting mesh?

**Comment by starki0815** [comment]
Here are pictures (the last one shows how it looks like if I change the code to generally use the manifold id to determine the new points on lines).
![plane_02](https://user-images.githubusercontent.com/45016248/53681247-881d9080-3cef-11e9-8ecf-09fb0559bfc0.png)
![plane_01](https://user-images.githubusercontent.com/45016248/53681248-881d9080-3cef-11e9-9c5f-228d909d1b8a.png)
![plane_patched](https://user-images.githubusercontent.com/45016248/53681249-8a7fea80-3cef-11e9-9300-e901b384c844.png)

I think what happens is the following: In the first refinement cycle, the manifold id of the cell (witch is numbers::flat_manifold_id) is used to decide which manifold to use for the midpoint. So, for this a flat manifold is used. Moreover, according to the above code, the boundary_id (which is 0) is used to decide which manifold to use for the new points on the lines. As no manifold is associated with manifold_id 0, also in this case a flat manifold is used. So nothing goes wrong in the first refinement cycle. However, in the second refinement cycle, there are lines to refine which are not at the boundary. So, according to the above code, the material_id of the adjacent cell (which is 1 in my example) is taken to decide which manifold to use for the new points on the interior lines. Because there is a spherical manifold associated with manifold_id 1, this manifold will be used for this purpose. This then leads to the distorted mesh.

Also, if I assign a boundary_id of 1 to the boundaries of the cell, already the first refinement cycle goes wrong, because then the boundary_id is used to determine the manifold for the new points on boundary lines:
![plane_boundary_id_1](https://user-images.githubusercontent.com/45016248/53681439-4b06cd80-3cf2-11e9-98b0-7a7640a4db54.png)

This all implies that rather unexpected behavior can occur if one uses a number for the material_id or boundary_id for which also a non-flat manifold object is attached to the triangulation (except if one explicitly defines manifold_ids!=numbers:flat_manifold_id for every single line).

**Comment by bangerth** [comment]
I haven't tried to debug this, but I thought we had gotten rid of boundary ids to determine manifolds at some (recent) point in history. Are you using the current developer version, or is this some older version of deal.II?

**Comment by masterleinad** [comment]
Fixed in #7775.
